Label Source feature

sudharsana-kjl opened the pull request:
This PR hopes to add a feature where we can include a label for the data source that can be used to add a version of the data set as well. This will help in having multiple versions of the dataset that can be trained.

sudharsana-kjl commented:
@pdxjohnny As per our discussion, I've tried implementing the label feature for JSONSource first. But the tests are failing.I've examined this behaviour by giving multiple print statements.

I tried examining the errors by first printing ```self.mem.values()``` in ```dump_fd()``` of ```JSONSource``` along with the type of each item in it. I get something like the following:
```
....................E<class 'list'>
[{}]
<class 'dffml.repo.Repo'>
feedface    (42.0% confidence) 0 classified as: 1
PetalLength                   3.9
PetalWidth                    1.2
SepalLength                   5.8
SepalWidth                    2.7
<class 'dffml.repo.Repo'>
Undetermined (0.0% confidence) 1 classified as: 1
PetalLength                   3.9
PetalWidth                    1.2
SepalLength                   5.8
SepalWidth                    2.7
E.<class 'list'>
[{}]
<class 'dffml.repo.Repo'>
Undetermined (0.0% confidence) 0.6425988695208351
<class 'dffml.repo.Repo'>
Undetermined (0.0% confidence) 0.6332119944551861
<class 'dffml.repo.Repo'>
Undetermined (0.0% confidence) 0.5969591376572548
<class 'dffml.repo.Repo'>
Undetermined (0.0% confidence) 0.4939760569239926
<class 'dffml.repo.Repo'>
Undetermined (0.0% confidence) 0.7380119951729546
<class 'dffml.repo.Repo'>
Undetermined (0.0% confidence) 0.595178964881829
<class 'dffml.repo.Repo'>
Undetermined (0.0% confidence) 0.9199616103174597
<class 'dffml.repo.Repo'>
Undetermined (0.0% confidence) 0.9820619507497166
<class 'dffml.repo.Repo'>
Undetermined (0.0% confidence) 0.8079311069896653
<class 'dffml.repo.Repo'>
Undetermined (0.0% confidence) 0.77215460583149
E<class 'list'>
[{}]
<class 'dffml.repo.Repo'>
Undetermined (0.0% confidence) 0.7294778597257038
```
But when i try to print the dictionary repo object, it doesnt work and the type of object itself changes to ```<class 'dict'>``` and throws an error like ```AttributeError: 'dict' object has no attribute 'dict'```

pdxjohnny commented on dffml/source/json.py:
```diff
@@ -20,13 +20,22 @@ class JSONSource(FileSource, MemorySource):
+                    for src_url, data in repos.items()
+                }
+            ]
         }
```
You're syntax here is incorrect `{ variable: [] }` is invalid. Only strings can go inside dictionary definitions. What you were doing is valid in JavaScript, but not python :)

You'll need to define this at the top

```python
class MissingLabel(Exception):
    pass
```

Then, `load_fd` should be along the lines of:

```python
        repos = json.load(fd)
        if not self.config.label in repos:
            raise MissingLabel('%r not found in %s', self.config.label, ', '.join(repos.keys()))
        self.mem = {
            src_url: Repo(src_url, data=data)
            for src_url, data in repos[self.config.label].items()
        }
```

pdxjohnny commented on tests/source/test_file.py:
```diff
@@ -79,29 +88,43 @@ def test_args(self):
             )
         )
         self.assertEqual(config.filename, "feedface")
         self.assertTrue(config.readonly)
```
Check that label is being set to `default-label` here

sudharsana-kjl replied:
Done :+1:

pdxjohnny commented on dffml/source/json.py:
```diff
@@ -20,13 +20,22 @@ class JSONSource(FileSource, MemorySource):
+                    {repo.src_url: repo.dict() for repo in self.mem.values()}
+                ]
+            },
+            fd,
```
You need to load the file with `json.load` then update the label within the loaded JSON, then dump the modified version back into the file. My guess would be that this is how it should be done:

```python
repos = json.load(fd)
repos[self.config.label] = {repo.src_url: repo.dict() for repo in self.mem.values()}
fd.seek(0)
json.dump(repos, fd)
```

pdxjohnny commented on dffml/source/file.py:
```diff
@@ -19,6 +19,7 @@
 
 class FileSourceConfig(BaseConfig, NamedTuple):
     filename: str
+    label: str
```
Give this a default value of `"unlabeled"`

sudharsana-kjl replied:
done :+1:

pdxjohnny commented on dffml/util/cli/cmds.py:
```diff
@@ -111,7 +111,8 @@ class SourcesCMD(CMD):
                         os.path.expanduser("~"), ".cache", "dffml.json"
-                    )
+                    ),
+                    label="default",
```
Anywhere where you've explicitly set label, remove it. Unless it's for a testcase to check that data with different labels is successfully living together in the same source. (Please add a testcase for this, so two JSONSource's with the same file and different labels, check that loading and saving them both results in data of both labels being present, repeat for CSVSource when that's ready).

sudharsana-kjl replied:
Okay :+1: I'll add the test cases once we debug the errors.

sudharsana-kjl commented on tests/source/test_file.py:
```diff
@@ -79,29 +88,43 @@ def test_args(self):
         )
         self.assertEqual(config.filename, "feedface")
+        self.assertEqual(config.label, "default-label")
         self.assertFalse(config.readonly)
```
Done :+1:

sudharsana-kjl commented on tests/source/test_file.py:
```diff
@@ -70,6 +70,15 @@ def test_args(self):
                                     "config": {},
                                 },
+                                "label": {"arg": Arg(), "config": {}},
+                                "readonly": {
```
Sorry :sweat_smile: My bad, I didn't notice this.

sudharsana-kjl commented on dffml/source/file.py:
```diff
@@ -114,11 +115,13 @@ def args(cls, args, *above) -> Dict[str, Arg]:
             "readonly",
             Arg(type=bool, action="store_true", default=False),
         )
+        cls.config_set(args, above, "label", Arg())
```
okay :+1:

sudharsana-kjl commented on CHANGELOG.md:
```diff
@@ -5,6 +5,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased]
+### Added
+- Added a feature to include labels for the dataset
```
Okay :+1:

pdxjohnny commented on tests/source/test_file.py:
```diff
@@ -82,19 +89,27 @@ def test_config_readonly_default(self):
 
-    def config(self, filename, readonly=False):
-        return FileSourceConfig(filename=filename, readonly=readonly)
+    def config(self, filename, label, readonly=False):
```
```python
def config(self, filename, label="unlabeled", readonly=False):
```

sudharsana-kjl replied:
Done :+1:

sudharsana-kjl commented on dffml/source/json.py:
```diff
@@ -19,14 +23,19 @@ class JSONSource(FileSource, MemorySource):
+        print("\n\nREPOSSS ")
+        print(repos)
+        if not self.config.label in repos.keys():
+            raise MissingLabel('%r not found in %s', self.config.label, ', '.join(repos.keys()))
```
oh okay.. :+1:

sudharsana-kjl commented:
Now, all the errors are gone while running tests. @pdxjohnny Could you please review the code and suggest any changes?

To debug ```JSONDecodeError``` I tried printing the file stream and also line by line, but i'm still not sure why it occurs only in certain test scenarios.  For now I've added ```try``` and ```except```.

pdxjohnny commented:
STYLE check can be fixed with

```console
black dffml tests examples
```

You'll need to `pip install black`

https://github.com/python/black

pdxjohnny commented:
Hold off on pushing anything for now. I'm about ot push to this branch

pdxjohnny commented:
@sudharsana-kjl I did some cleanup of file tests and then fixed whatever it was that was weird that happened with Git where it wouldn't let me rebase.

Here's the commit that adds labels, I changed what you had slightly https://github.com/intel/dffml/pull/80/commits/8666f9581fdcbedb8f46fcad83a7a1ef5e876ecf

Mainly I decided it makes the most sense to hold everything in memory `self.repos` and then not worry about if a label is missing, because if we open a source and try to merge then it's going to end up not having it to begin with. Which, now as I write this I realize that the `merge` CLI command won't work right now because of the `self.repos` approach. We'll talk about this in the meeting on thursday

This is the test for labels: https://github.com/intel/dffml/pull/80/commits/8f6953ef4c443e3dcff5ca312b9efcbf446e4f38

Please review and let me know if I missed anything or you have any questions about why I did something :)

sudharsana-kjl commented on dffml/util/testing/source.py:
```diff
@@ -60,40 +59,78 @@ class SourceTest(abc.ABC):
+
+class FileSourceTest(SourceTest):
+    """
+    Test case class used to test a Source implementation. Subclass from and set
```
I guess you've missed out "Test case class used to test a Source implementation. Subclass from ___ and set..."

pdxjohnny commented on dffml/util/testing/source.py:
```diff
@@ -60,40 +59,78 @@ class SourceTest(abc.ABC):
+
+    async def test_label(self):
+        with tempfile.TemporaryDirectory() as testdir:
+            self.testfile = os.path.join(testdir, str(random.random))
```
Need to call `random.random()`

pdxjohnny commented:
```console
git checkout master
git branch -D label_and_version_feature
git fetch origin pull/80/head:label_and_version_feature
git checkout label_and_version_feature
```

pdxjohnny commented:
Don't forget: You'll need re-add the fd.seek(0) then before you json.dump() also add a fd.truncate(0)

pdxjohnny commented on dffml/source/json.py:
```diff
@@ -17,16 +17,21 @@ class JSONSource(FileSource, MemorySource):
-        json.dump(
-            {repo.src_url: repo.dict() for repo in self.mem.values()}, fd
-        )
+        self.repos[self.config.label] = {
```
```python
repos = json.load(fd)
repos[self.config.label] = { ... }
fd.seek(0)
# Delete contents of file
fd.truncate(0)
json.dump(repos, fd)
```

pdxjohnny replied:
Also, remove self.repos (so remove the `__init__` method please too)

pdxjohnny commented on dffml/source/json.py:
```diff
@@ -17,16 +17,21 @@ class JSONSource(FileSource, MemorySource):
         self.mem = {
             src_url: Repo(src_url, data=data)
-            for src_url, data in repos.items()
+            for src_url, data in self.repos.get(self.config.label, {}).items()
```
```python
for src_url, data in self.repos.get(self.config.label, {}).items()
```

becomes

```python
for src_url, data in json.load(fd).get(self.config.label, {}).items()
```

pdxjohnny commented on dffml/util/testing/source.py:
```diff
@@ -60,40 +59,78 @@ class SourceTest(abc.ABC):
+            async with labeled:
+                async with labeled() as lctx:
+                    repo = await lctx.repo("0")
+                    self.assertNotIn("life", repo.features())
```
Add something which opens both at the same time then closes, then opens again and verifiy's they are still there.

```python
            async with unlabeled, labeled:
                async with unlabeled() as uctx, labeled() as lctx:
                    await uctx.update(
                        Repo("0", data={"features": {"feed": 1}})
                    )
                    await lctx.update(
                        Repo("0", data={"features": {"face": 2}})
                    )
            async with unlabeled, labeled:
                async with unlabeled() as uctx, labeled() as lctx:
                    repo = await uctx.repo("0")
                    self.assertNotIn("feed", repo.features())
                    repo = await lctx.repo("0")
                    self.assertNotIn("face", repo.features())
```

do this first, it shouldn't pass, change the code, it should pass

sudharsana-kjl commented:
I've changed self.repos to just repos in JSONSource. Hopefully merge command works now. 
The problem happened when we tried to do ```repos = json.load(fd)``` in ```dump_fd()``` for an empty stream. So I've added extra conditions to avoid this. The empty stream is passed in certain tests like test_label, etc. where there's not __aenter__() called but __aexit__() is called and a new tmp file is generated for dumping.

pdxjohnny commented:
@sudharsana-kjl  I wrote the test for Merge, and there are issues with some of the CLI stuff, I'm going to keep pushing to this and rebase it onto master as soon as things are ready. You should do you're work on CSVSource in another branch, just FYI.

pdxjohnny commented:
I'm going to merge this, even though `Merge` is broken, so that you can get a start on CSV labels if you want to. See my comments on gitter with regards to the other PR which was reverted
